k8s/admission.go

```diff
@@ -0,0 +1,232 @@
+		})
+		// TODO: unsure on this
+		if len(request.Object.CommonMetadata().Labels) == 0 {
+			ops := append(make([]resource.PatchOperation, 0), resource.PatchOperation{
+				Path:      "/metadata/labels",
+				Operation: resource.PatchOpAdd,
+				Value: map[string]string{
+					versionLabel: request.Version,
+				},
+			})
+			ops = append(ops, resp.PatchOperations...)
+			resp.PatchOperations = ops
+		} else {
+			resp.PatchOperations = append(resp.PatchOperations, resource.PatchOperation{
+				Path:      "/metadata/labels/" + versionLabel, // Set the internal version label to the version of the endpoint
+				Operation: resource.PatchOpAdd,
+				Value:     request.Version,
+			})
+		}
+	case resource.AdmissionActionUpdate:
+		// Patch is tricky when it comes to add vs replace operations for maps, like labels and annotations
+		resp.PatchOperations = append(resp.PatchOperations, resource.PatchOperation{
+			Path:      "/metadata/updatedBy", // Set createdBy to the request user
+			Operation: resource.PatchOpReplace,
+			Value:     request.UserInfo.Username,
+		}, resource.PatchOperation{
+			Path:      "/metadata/updateTimestamp", // Set the updateTimestamp to the creationTimestamp
+			Operation: resource.PatchOpReplace,
+			Value:     now().Format(time.RFC3339Nano),
+		})
+		// TODO: unsure on this
+		if len(request.Object.CommonMetadata().Labels) == 0 {
+			ops := append(make([]resource.PatchOperation, 0), resource.PatchOperation{
+				Path:      "/metadata/labels",
+				Operation: resource.PatchOpAdd,
+				Value: map[string]string{
+					versionLabel: request.Version,
+				},
+			})
+			ops = append(ops, resp.PatchOperations...)
+			resp.PatchOperations = ops
+		} else {
+			resp.PatchOperations = append(resp.PatchOperations, resource.PatchOperation{
+				Path:      "/metadata/labels/" + versionLabel, // Set the internal version label to the version of the endpoint
+				Operation: resource.PatchOpAdd,
+				Value:     request.Version,
+			})
+		}
+	default:
+		// Do nothing
+	}
+	return resp, nil
+}
+
+// NewOpinionatedMutatingAdmissionController creates a pointer to a new OpinionatedMutatingAdmissionController wrapping the
+// provided mutateFunc (nil mutateFunc argument is allowed, and will cause the controller to not call the underlying function)
+func NewOpinionatedMutatingAdmissionController(mutateFunc func(request *resource.AdmissionRequest) (*resource.MutatingResponse, error)) *OpinionatedMutatingAdmissionController {
+	return &OpinionatedMutatingAdmissionController{
+		MutateFunc: mutateFunc,
+	}
+}
+
+// OpinionatedValidatingAdmissionController implements resource.ValidatingAdmissionController and performs initial
+// validation on reserved metadata fields which are stores as annotations in kubernetes, ensuring that if any changes are made,
+// they are allowed, before calling the underlying admission validate function.
+type OpinionatedValidatingAdmissionController struct {
+	ValidateFunc func(*resource.AdmissionRequest) error
```

Comment:
Same here – IMHO wrapping the interface makes more sense, that way we can have consistent patter of nesting the controllers.